crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3936,6 +3940,55 @@ impl<'db> TypeInferenceBuilder<'db> {
         ))
     }
 
+    /// Returns the symbol of the first parameter if the scope is function-like (function or lambda).
+    /// Returns `None` if the scope is not function-like, or if there are no parameters.
+    fn first_param_symbol_in_scope(&self, scope: ScopeId) -> Option<Symbol<'db>> {
+        let file_scope_id = scope.file_scope_id(self.db());
+        let symbol_table = self.index.symbol_table(file_scope_id);
+        let use_def = self.index.use_def_map(file_scope_id);
+
+        let first_param_name = match scope.node(self.db()) {
+            NodeWithScopeKind::Function(f) => f.parameters.iter().next().map(|p| p.name().as_str()),
+            NodeWithScopeKind::Lambda(l) => l
+                .parameters
+                .as_ref()?
+                .args
+                .first()
+                .map(|p| p.name().as_str()),
+            _ => None,
+        }?;
+
+        let scoped_symbol = symbol_table.symbol_id_by_name(first_param_name)?;
+        let bindings = use_def.public_bindings(scoped_symbol);
+        Some(symbol_from_bindings(self.db(), bindings))
+    }
+
+    /// Returns the symbol of the nearest enclosing class for the given scope.
+    ///
+    /// This function walks up the ancestor scopes starting from the given scope,
+    /// and finds the closest class definition.
+    ///
+    /// Returns `None` if no enclosing class is found.
+    fn enclosing_class_symbol(&self, scope: ScopeId) -> Option<Symbol<'db>> {
+        self.index
+            .ancestor_scopes(scope.file_scope_id(self.db()))
+            .find_map(|(_, ancestor_scope)| {
+                if let NodeWithScopeKind::Class(class) = ancestor_scope.node() {
+                    // The class name is resolved in its parent scope
+                    let parent_scope_id = ancestor_scope.parent()?;
+                    let parent_symbol_table = self.index.symbol_table(parent_scope_id);
+                    let parent_use_def = self.index.use_def_map(parent_scope_id);
+
+                    let scoped_symbol =
+                        parent_symbol_table.symbol_id_by_name(class.name.as_str())?;
+                    let bindings = parent_use_def.public_bindings(scoped_symbol);
+                    Some(symbol_from_bindings(self.db(), bindings))
```

Comment:
Resolving the class symbol from `public_bindings` has a similar problem as above: if we have a class `A` in the outer scope, and later in that scope the name `A` is reassigned to something else, a use of `super` inside a method of `A` will find the wrong binding of `A` in the outer scope.

What we need to do here instead is instead get the class node from the outer scope, get its Definition from the semantic index, and use `infer_definition_types` query (see `infer_definition` method for an example), then pull out the type for the definition from the resulting `TypeInference`.

This function should also just return a `Type`, not a `Symbol`, as boundness is not relevant here: the class statement node definitely has a type.